third_party/move/move-binary-format/src/file_format.rs

```diff
@@ -2934,6 +2969,134 @@ pub enum Bytecode {
+        stack.
+
+        Notice that the type as part of this instruction is redundant for
+        execution semantics. Since the closure is expected to be on top of the stack,
+        it can decode the arguments underneath without type information.
+        However, the type is required for the current implementation of
+        static bytecode verification.
+    "#]
+    #[static_operands = "[u8_value]"]
+    #[semantics = r#"
+        stack >> function_handle
+
+        let [func, k, [arg_0, .., arg_{k-1}]] = function_handle
+        // Information like the function signature are loaded from the file format
+        i = u8_value
+        n = func.num_params
+        if i + k > n then abort
+
+        stack >> arg'_{i-1}
+        ..
+        stack >> arg'_0
+        new_function_handle = [func, k + i, [arg_0, .., arg_{k-1}, arg'_0, .., arg'_{i-1}]]
+        stack << new_function_handle
+    "#]
+    #[runtime_check_epilogue = r#"
+        // NOT: assert func visibility rules
+        func param types match provided parameter types
+    "#]
+    #[gas_type_creation_tier_0 = "closure_ty"]
+    EarlyBindFunction(SignatureIndex, u8),
+
+    #[group = "closure"]
+    #[description = r#"
+        `InvokeFunction(|t1..tn|r with a)` calls a function value of the specified type,
+        with `n` argument values from the stack.
+
+        On top of the stack is the closure being evaluated, underneath the arguments:
+        `[c,vn,..,v1] + stack`. The type of the closure must match the type specified in
+        the instruction, with abilities `a` a subset of the abilities of the closure value.
+        A value `vi` on the stack must have type `ti`.
+
+        Notice that the type as part of this instruction is redundant for
+        execution semantics. Since the closure is expected to be on top of the stack,
+        it can decode the arguments underneath without type information.
+        However, the type is required for the current implementation of
+        static bytecode verification.
+
+        The arguments are consumed and pushed to the locals of the function.
+
+        Return values are pushed onto the stack from the first to the last and
+        available to the caller after returning from the callee.
+
+        During execution of the function conservative_invocation_mode is enabled;
+        afterwards, it is restored to its original state.
+    "#]
+    #[semantics = r#"
+        stack >> function_handle
+        let [func, k, [arg_0, .., arg_{k-1}]] = function_handle
+        // Information like the function signature are loaded from the file format
+        // and compared with the signature parameter.
+        n = func.num_params
+        ty_args = if func.is_generic then func.ty_args else []
+
+        n = func.num_params
+        stack >> arg_{n-1}
+        ..
+        stack >> arg_k
+
+        old_conservative_invocation_mode = conservative_invocation_mode
```

Comment:
For LAMBDA MVP, we are playing it very safe and disallowing any call to module `m` if there is any dynamic call on the stack.  This is enforced by a VM state flag `conservative_invocation_mode`, which is set if we are in a dynamic call.

This is because dynamic calls violate the static dependency guarantees.  After we have the MVP implemented we will see how far we can prove dynamic safety, e.g., using access specifiers.

There was an error here in that the dynamic call also needs to always check that its module is not already on the stack.  Fixed.